freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -16,171 +16,188 @@
-            if (isModuleFS(cs)) fromClass(cs) else List(meth.typeConstructor)
-          case _ => Nil
-        }
+  private[this] lazy val wildcard =
+    TypeDef(Modifiers(Flag.PARAM), typeNames.WILDCARD, List(),
+      TypeBoundsTree(EmptyTree, EmptyTree))
 
-      fromClass(s)
-    }
+  private[this] lazy val FFtypeConstructor =
+    TypeDef(Modifiers(Flag.PARAM), FF, List(wildcard),
+      TypeBoundsTree(EmptyTree, EmptyTree))
 
-    val AA = freshTypeName("AA$")
-
-    def mkCoproduct(algebras: List[Type]): List[TypeDef] =
-      algebras.map(x => TermName(x.toString) ) match {
-        case Nil => q"type Op[$AA] = Nothing" :: Nil
-
-        case alg :: Nil => q"type Op[$AA] = $alg.Op[$AA]" :: Nil
-
-        case alg0 :: alg1 :: algs =>
-          /* We create several type aliases, where the type names are generated fresh names,
-           * C0, C1, ..., C{n-1}, where n is the length of algebras. We then generate aliases:
-           *
-           * type C1 = A1 |+| A0
-           * type C2 = A2 |+| C1
-           * type Ci = Ai |+| C{i-1}
-           * type C{n-1} = A{n-1} |+| C{n-2
-           * type Op{n-1}= C{n-1}
-           */
-          val ccs: List[TypeName] = algebras.map( _ => freshTypeName("CC$") )
-          val tyDef1 = q"type ${ccs(1)}[$AA] = Coproduct[$alg1.Op, $alg0.Op, $AA]"
-          val tyDefs = algebras.zipWithIndex.drop(2).map { case (alg, pos) =>
-            q"type ${ccs(pos)}[$AA] = Coproduct[$alg.Op, ${ccs(pos-1)}, $AA]"
-          }
-          val opDef = q"type Op[$AA] = ${ccs.last}[$AA]"
-          tyDef1 :: (tyDefs :+ opDef)
-      }
+  private[this] implicit final class ValidateFlatmap[E, A](v: Validated[E, A]) {
+    def flatMap[EE >: E, B](f: A => Validated[EE, B]): Validated[EE, B] =
+      v.andThen(f)
+  }
 
-    // findAlgebras starts from the ClassSymbol of the `@moudle-annotated trait
-    val algebras   = findAlgebras(weakTypeOf[A].typeSymbol.companion.asClass)
-    val coproducts = mkCoproduct(algebras)
-    //ugly hack because, as String,  it does not typecheck, early which we need for types to be in scope
-    val parsed     = coproducts.map( cop => c.parse(cop.toString))
-
-    val expr = if (algebras.length >= 2)
-      q"""new {
-        import _root_.cats.data.Coproduct
-        ..$parsed
-      }"""
+  def impl(annottees: c.Expr[Any]*): Tree =
+    foldAbort(for {
+      cls             <- decodeAnnotationTarget(annottees.toList)
+      effectTuples    <- gatherEffects(cls)
+      classTree       <- makeClassTree(cls)
+      objectTree      <- makeObjectTree(cls, effectTuples)
+    } yield q"$classTree; $objectTree")
+
+  def computeCoproductImpl: Tree =
+    foldAbort(for {
+      tpe             <- compassionateCompanionTypeOf(c.internal.enclosingOwner.owner)
+      fullEffectTypes =  expandEffectType(tpe)
+      fullEffectTrees <- fullEffectTypes.traverse(typeToTypeTree)
+      //coproductTree   <- makeCatsCoproduct(fullEffectTrees) // TODO: make this configurable?
```

Comment:
Should I remove this line? I thought it would be good to keep around for when we feature flag targeting both iota and cats.